File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTransaction.java
package nl.tudelft.ewi.ds.bankchain.bank.bunq;

import java.util.Currency;
import java.util.Date;
import java.util.List;

import java8.util.concurrent.CompletableFuture;
import nl.tudelft.ewi.ds.bankchain.bank.Account;
import nl.tudelft.ewi.ds.bankchain.bank.Transaction;
import nl.tudelft.ewi.ds.bankchain.bank.bunq.api.PaymentService;

/**
 * A transaction from the Bunq bank
 *
 * @author Jos Kuijpers
 */
public class BunqTransaction implements Transaction {


    private int id;
    private Date date;
    private float value;
    private String description;
    private Currency currency;
    private Account counterAccount;
    private Account account;


    public BunqTransaction(PaymentService.ListResponse.Payment payment, Account account) {
        date = new Date(22111990); //payment.getCreationDate(); Todo fix payment
        value = payment.amount.getValue();
        currency = payment.amount.getCurrency();
        description = payment.description;
        this.account = account;
        this.id = payment.id;
        this.counterAccount = new BunqAccount(payment.counterParty.iban, -1, new BunqParty(payment.counterParty.name, -1));
    }

    /**
     * used to create an outgoing trasnaction
     * @param amount should be negative
     * @param account source of the money
     * @param counterAccount recepient
     * @param currency
     * @param description
     */
    public BunqTransaction(float amount, Account account, Account counterAccount,Currency currency, String description){
        date = getDate();
        id = -1;
        this.currency = currency;
        value = amount;
        this.account = account;
        this.counterAccount = counterAccount;
        this.description = description;
    }


    @Override
    public Date getDate() {
        return date;
    }

    @Override
    public Float getValue() {
        return value;
    }

    @Override
    public Currency getCurrency() {
        return currency;
    }

    @Override
    public Account getCounterAccount() {
        return counterAccount;
    }

    @Override
    public Account getAcount() {
        return account;
    }


    @Override
    public String getDescription() {
        return description;
    }

    @Override
    public int getId() {
        return id;
    }

    public void setId(int id){
        this.id = id;
    }

    @Override
    public String toString() {
        return "<Payment desc:'" + getDescription() + "'>";
    }

    public PaymentService.PostRequest convertToRequest(){
        PaymentService.PostRequest p = new PaymentService.PostRequest();
        PaymentService.Amount a = new PaymentService.Amount();
        PaymentService.Counterparty_alias c = new PaymentService.Counterparty_alias();
        c.name = counterAccount.getParty().getName();
        c.value = counterAccount.getIban();
        a.currency = "EUR";
        a.value = Float.toString(value);
        p.description = description;
        return p;
    }
}

File: BankChain/bankver/src/main/java/nl/tudelft/ewi/ds/bankver/bank/bunq/BunqTransaction.kt
package nl.tudelft.ewi.ds.bankver.bank.bunq

import java.util.Currency
import java.util.Date

import nl.tudelft.ewi.ds.bankver.bank.Account
import nl.tudelft.ewi.ds.bankver.bank.Transaction
import nl.tudelft.ewi.ds.bankver.bank.bunq.api.PaymentService

/**
 * A transaction from the Bunq bank
 */
class BunqTransaction : Transaction {

    override var id: Int
    override val date: Date
    override val value: Float
    override val description: String
    override val currency: Currency
    override val account: Account
    override val counterAccount: Account


    constructor(payment: PaymentService.ListResponse.Payment, account: Account) {
        this.id = payment.id
        this.date = Date(22111990) //payment.getCreationDate(); TODO fix payment
        this.value = payment.amount.value!!
        this.description = payment.description
        this.currency = payment.amount.currency
        this.account = account
        this.counterAccount = BunqAccount(payment.counterParty.iban, -1, BunqParty(payment.counterParty.name, -1))
    }

    /**
     * used to create an outgoing transaction
     * @param amount
     * @param account        source of the money
     * @param counterAccount recipient
     * @param currency
     * @param description
     */
    constructor(amount: Float, account: Account, counterAccount: Account, currency: Currency, description: String) {
        this.id = -1
        this.date = Date(0)
        this.value = amount
        this.description = description
        this.currency = currency
        this.account = account
        this.counterAccount = counterAccount
    }

    fun convertToRequest(): PaymentService.PostRequest {
        val p = PaymentService.PostRequest()
        val a = PaymentService.Amount()
        val c = PaymentService.CounterpartyAlias()
        c.name = counterAccount.party.name
        c.value = counterAccount.iban
        a.currency = "EUR"
        a.value = java.lang.Float.toString(Math.abs(value)) //usually the outgoing amount should be negative this is the exception so it's covered here
        p.description = description
        p.amount = a
        p.counterparty_alias = c
        return p
    }
}
